syntax = "proto3";

package post;

message PostData {
    string title = 2;
    string description = 3;
    bool is_private = 7;
    repeated string tags = 8;
}

message Post {
    int32 id = 1;
    PostData data = 2;
    int32 creator_id = 3;
    string created_at = 4;
    string updated_at = 5;
}

message CreatePostRequest {
    PostData post_data = 1;

    int32 creator_id = 1000;
}

message DeletePostRequest {
    int32 post_id = 1;

    int32 creator_id = 1000;
}

message UpdatePostRequest {
    int32 post_id = 1;
    PostData post_data = 2;

    int32 creator_id = 1000;
}

message GetPostRequest {
    int32 post_id = 1;

    int32 creator_id = 1000;
}

message ListPostsRequest {
    int32 creator_id = 1000;
}

message PostResponse {
    Post post = 1;

    string error = 1000;
}

message DeletePostResponse {
    string error = 1000;
}

message ListPostsResponse {
    repeated Post posts = 1;
    
    string error = 1000;
}

message Comment {
    int32 id = 1;
    int32 post_id = 2;
    int32 creator_id = 3;
    string text = 4;
    string created_at = 5;
}

message CreateCommentRequest {
    int32 creator_id = 1;
    int32 post_id = 2;
    string text = 3;
}

message CreateCommentResponse {
    Comment comment = 1;
    string error = 1000;
}

message ListCommentsRequest {
    int32 creator_id = 1;
    int32 post_id = 2;
    int32 page = 3;
    int32 page_size = 4;
}

message ListCommentsResponse {
    repeated Comment comments = 1;
    string error = 1000;
}

service PostService {
    rpc CreatePost(CreatePostRequest) returns (PostResponse);
    rpc DeletePost(DeletePostRequest) returns (DeletePostResponse);
    rpc UpdatePost(UpdatePostRequest) returns (PostResponse);
    rpc GetPost(GetPostRequest) returns (PostResponse);
    rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);
    
    rpc CreateComment(CreateCommentRequest) returns (CreateCommentResponse);
    rpc ListComments(ListCommentsRequest) returns (ListCommentsResponse);
}
